cirq-core/cirq/devices/superconducting_qubits_noise_properties.py

```diff
@@ -0,0 +1,200 @@
+        readout_errors: Dict[cirq.Qid, np.ndarray] of qubits to their readout
+            errors in matrix form: [P(read |1> from |0>), P(read |0> from |1>)].
+        gate_pauli_errors: dict of OpIdentifiers (a gate and the qubits it
+            targets) to the Pauli error for that operation. Keys in this dict
+            must have defined qubits.
+        validate: If True, verifies that t1 and tphi qubits sets match, and
+            that all symmetric two-qubit gates have errors which are
+            symmetric over the qubits they affect. Defaults to True.
+    """
+
+    gate_times_ns: Dict[type, float]
+    t1_ns: Dict['cirq.Qid', float]
+    tphi_ns: Dict['cirq.Qid', float]
+    readout_errors: Dict['cirq.Qid', List[float]]
+    gate_pauli_errors: Dict[OpIdentifier, float]
+
+    validate: bool = True
+    _qubits: List['cirq.Qid'] = field(init=False, default_factory=list)
+
+    def __post_init__(self):
+        if not self.validate:
+            return
+        t1_qubits = set(self.t1_ns)
+        tphi_qubits = set(self.tphi_ns)
+
+        if t1_qubits != tphi_qubits:
+            raise ValueError('Keys specified for T1 and Tphi are not identical.')
+
+        # validate two qubit gate errors.
+        self._validate_symmetric_errors('gate_pauli_errors')
+
+    def _validate_symmetric_errors(self, field_name: str) -> None:
+        gate_error_dict = getattr(self, field_name)
+        for op_id in gate_error_dict:
+            if len(op_id.qubits) != 2:
+                # single qubit op_ids also present, or generic values are
+                # specified. Skip these cases
+                if len(op_id.qubits) > 2:
+                    raise ValueError(
+                        f'Found gate {op_id.gate_type} with {len(op_id.qubits)} qubits. '
+                        'Symmetric errors can only apply to 2-qubit gates.'
+                    )
+            elif op_id.gate_type in self.symmetric_two_qubit_gates():
+                op_id_swapped = OpIdentifier(op_id.gate_type, *op_id.qubits[::-1])
+                if op_id_swapped not in gate_error_dict:
+                    raise ValueError(
+                        f'Operation {op_id} of field {field_name} has errors '
+                        f'but its symmetric id {op_id_swapped} does not.'
+                    )
+            elif op_id.gate_type not in self.asymmetric_two_qubit_gates():
+                # Asymmetric gates do not require validation.
+                raise ValueError(
+                    f'Found gate {op_id.gate_type} which does not appear in the '
+                    'symmetric or asymmetric gate sets.'
+                )
+
+    @property
+    def qubits(self) -> List['cirq.Qid']:
+        """Qubits for which we have data"""
+        if not self._qubits:
+            self._qubits = sorted(self.t1_ns)
+        return self._qubits
+
+    @classmethod
+    @abc.abstractmethod
+    def single_qubit_gates(cls) -> Set[type]:
```

Comment:
Type hints are kind of loose here, could we convert this and below to \'cirq.Gate\' ?